pkg/policy/mapstate.go

```diff
@@ -220,32 +244,176 @@ func getNets(identities Identities, ident uint32) []*net.IPNet {
+		} else {
+			ms.allows[k] = v
+		}
+	}
+}
+
+// Delete removes the Key an related MapStateEntry.
+func (ms *mapState) Delete(k Key) {
+	if ms != nil {
+		delete(ms.allows, k)
+		delete(ms.denies, k)
+	}
+}
+
+// ForEach iterates over every Key MapStateEntry and stops when the function
+// argument returns false. It returns false iff the iteration was cut short.
+func (ms *mapState) ForEach(f func(Key, MapStateEntry) (cont bool)) (complete bool) {
+	if ms == nil {
+		return true
+	}
+
+	if complete := ms.ForEachAllow(f); !complete {
+		return complete
+	}
+
+	return ms.ForEachDeny(f)
+}
+
+// ForEachAllow iterates over every Key MapStateEntry that isn't a deny and
+// stops when the function argument returns false
+func (ms *mapState) ForEachAllow(f func(Key, MapStateEntry) (cont bool)) (complete bool) {
+	if ms == nil {
+		return true
+	}
+
+	for k, v := range ms.allows {
+		if !f(k, v) {
+			return false
+		}
+	}
+	return true
+}
+
+// ForEachDeny iterates over every Key MapStateEntry that is a deny and
+// stops when the function argument returns false
+func (ms *mapState) ForEachDeny(f func(Key, MapStateEntry) (cont bool)) (complete bool) {
+	if ms == nil {
+		return
+	}
+
+	for k, v := range ms.denies {
+		if !f(k, v) {
+			return false
+		}
+	}
+	return true
+}
+
+// Len returns the length of the map
+func (ms *mapState) Len() int {
+	return len(ms.allows) + len(ms.denies)
+}
+
+// Equals determines if this MapState is equal to the
+// argument MapState
+func (msA *mapState) Equals(msB MapState) bool {
+	if (msA == nil) != (msB == nil) {
+		return false
+	} else if msA == nil && msB == nil {
+		return true
+	} else if msA.Len() != msB.Len() {
+		return false
+	}
+
+	return msB.ForEach(func(kA Key, vA MapStateEntry) bool {
+		if vB, ok := msB.Get(kA); ok {
+			if !(&vB).DatapathEqual(&vA) {
+				return false
+			}
+		} else {
+			return false
+		}
+
+		return true
+	})
+}
+
 // AddDependent adds 'key' to the set of dependent keys.
-func (keys MapState) AddDependent(owner Key, dependent Key, changes ChangeState) {
-	if e, exists := keys[owner]; exists {
-		keys.addDependentOnEntry(owner, e, dependent, changes)
+func (ms *mapState) AddDependent(owner Key, dependent Key, changes ChangeState) {
+	if e, exists := ms.allows[owner]; exists {
+		ms.addDependentOnEntry(owner, e, dependent, changes)
+	} else if e, exists := ms.denies[owner]; exists {
+		ms.addDependentOnEntry(owner, e, dependent, changes)
 	}
 }
 
 // addDependentOnEntry adds 'dependent' to the set of dependent keys of 'e'.
-func (keys MapState) addDependentOnEntry(owner Key, e MapStateEntry, dependent Key, changes ChangeState) {
+func (ms *mapState) addDependentOnEntry(owner Key, e MapStateEntry, dependent Key, changes ChangeState) {
 	if _, exists := e.dependents[dependent]; !exists {
 		if changes.Old != nil {
-			changes.Old[owner] = e
+			changes.Old.Insert(owner, e)
 		}
 		e.AddDependent(dependent)
-		keys[owner] = e
+		if e.IsDeny {
+			ms.denies[owner] = e
```

Comment:
We probably should delete `owner` from the other map to make sure we don't end up 
with duplicate keys.